akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.stream.stage._
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+import scala.concurrent.duration._
+
+/**
+ * This is a more scalable and general case of [[GroupedWeightedWithin]]
+ * which groups a stream into vectors based on custom weight and time.
+ * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+ * That won't work for the use case of writing large files.
+ * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+ * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+ * This custom flow uses custom aggregator to support such use cases.
+ * Upstream inputs are continuously aggregated as they arrive.
+ * The aggregator/grouping can be terminated and emitted based on custom coditions
+ *
+ * @param seed        initiate the aggregated output with first input
+ * @param aggregate   sequentially aggregate input
+ * @param emitOnAgg   decide whether the current aggregator can be emitted, invoked after each aggregate
+ * @param emitOnTimer decide whether the current aggregator can be emitted, invoked after each timer event
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage
+ *                    there can be extra time between harvest and next stage receiving the emitted output
+ *                    time sensitive operations can be added here, such as closing an output channel
+ */
+class FoldWith[In, Agg, Out](
+    seed: In => Agg,
+    aggregate: (Agg, In) => Agg,
+    emitOnAgg: Agg => Boolean,
+    harvest: Agg => Out,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)] = None,
+    pullOnStart: Boolean = true)
+    extends GraphStage[FlowShape[In, Out]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      // mutable state to keep track of the aggregator status to coordinate the flow
+      // input/output handler callbacks are guaranteed to execute without concurrency
+      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
```

Comment:
We can expect devs looking at the operator internals to know this about the stream APIs, there's a lot of more places, if not every operator impl, you'd have to put this comment if not, so remove this.